lib/app_profiler.rb

```diff
@@ -79,25 +78,41 @@ def running?
     end
 
     def profiler
-      if @backend && !@backend.is_a?(profiler_backend)
-        raise ConfigurationError if @backend.running?
-
-        clear
-      end
-
-      @backend ||= profiler_backend.new
+      @backend ||= backend.new
     end
 
-    def backend=(backend)
-      return true if @backend&.is_a?(backend)
+    def backend=(new_backend)
+      return true if backend&.is_a?(new_backend)
 
       if running?
         raise ArgumentError,
-          "cannot change backend to #{backend::NAME} while #{profiler_backend::NAME} backend is running"
+          "cannot change backend to #{new_backend::NAME} while #{backend::NAME} backend is running"
       end
 
       clear
-      self.profiler_backend = backend
+      @profiler_backend = new_backend
+    end
+
+    def backend
+      @profiler_backend ||= DefaultBackend
+    end
+
+    def with_backend(backend)
+      orig_backend = AppProfiler.backend
+
+      if backend
+        if defined?(AppProfiler::VernierBackend) &&
+            backend == AppProfiler::VernierBackend::NAME
+          return yield unless (AppProfiler.backend = AppProfiler::VernierBackend)
```

Comment:
`backend=` raises when it is `running?` I think we need to either return `nil` or rescue, I prefer the former. 

Also the way `with_backend` is used in the middleware:

```
 AppProfiler.with_backend(params.backend) do
        AppProfiler.run(params_hash) do
         ......
```

...`return yield unless...` means that we will end up running `AppProfiler.run`? I think with_backend can just return `nil` or the actual backend, and then we can handle it in the middleware.

Reply:
> backend= raises when it is running? I think we need to either return nil or rescue, I prefer the former.

Good call, i missed this

> I think with_backend can just return nil or the actual backend, and then we can handle it in the middleware.

I'm a bit confused by this, how would it help to expose the backend back to the calling layer? And in the middleware, how would we get access to the profile when calling `AppProfiller.run` inside the `with_backend` block? Normally i think when wrapping using a `with_xxx` you would just return the value of the block 🤔